src/config.rs

```diff
@@ -1,118 +1,174 @@
-use arrayref::array_ref;
-use ed25519_dalek::*;
+// 
+// Manages creation, serialization (storage) and deserialization (loading) of HoloPort Device Seed
+// and Admin keys.
+// 
+// The HoloPort's /var/lib/holo-config.json Config is loaded, and the Config::V1.seed is used to
+// derive all Holochain, ZeroTier, etc. keypairs required for operation.  
+//
+// Admin Keypair
+// 
+// During creation of the Config, the Admin passphrase is known, and an Admin keypair is derived
+// from the entropy collected from the admin email "salt", the holochain Signing public key (Agent
+// ID) "pepper", and the Admin passphrase "password".  This yields an HoloPort Admin request signing
+// keypair, which is unique to each specific Holoport (due to the email "salt" and Agent ID
+// "pepper"), and is secured by the private Admin passphrase.  It can be easily derived (eg. in the
+// Admin UI), by knowing the Admin email (eg "admin@example.com"), the target Holoport Agent ID
+// (eg. "HcScjabc...123"), and the Admin passphrase (collected from the Admin via the UI).
+// 
+// We use a simple set of DPKI-compatible APIs to load, parse, validate, save and generate the needed
+// Seeds, etc.
+// 
+// Config Seed Security
+// 
+// Typical operation of the HoloPort requires unencrypted access to the seed (stored as 24 BIP39
+// words containing the 256-bit seed in plaintext), to support unattended rebooting.  This of course
+// requires physical and logical security of the removable media containing the serialized Config!
+// 
+// For those desiring additional security, you can opt to store your HoloPort seed *encrypted*
+// (stored as 48 BIP39 words containing the 256-bit seed ciphertext, plus 256-bit Authentication MAC
+// and random salt); of course, this requires the HoloPort Admin to log into the HoloPort on every
+// reboot, and enter the seed's passphrase to unlock it, and allow the HoloPort to continue booting.
+//
+use std::fmt;
 use failure::Error;
-use rand::{rngs::OsRng, Rng};
-use serde::*;
+use hcid;
+use rand::{
+    rngs::OsRng, Rng
+};
+use serde::{
+    Deserialize, Deserializer, Serialize, Serializer, de
+};
 
-use holochain_dpki::SEED_SIZE;
+use crate::dpki:: {
+    SEED_SIZE,
+    Seed, EncryptedSeed, SeedData,
+    SigningPublicKey, SigningSecretKey,
+    email_password_to_seed, signing_keypair_from_seed,
+};
 
-fn public_key_from_base64<'de, D>(deserializer: D) -> Result<PublicKey, D::Error>
-where
-    D: Deserializer<'de>,
-{
-    String::deserialize(deserializer)
-        .and_then(|s| {
-            base64::decode_config(&s, base64::STANDARD_NO_PAD)
-                .map_err(|err| de::Error::custom(err.to_string()))
-        })
-        .map(|bytes| PublicKey::from_bytes(&bytes))
-        .and_then(|maybe_key| maybe_key.map_err(|err| de::Error::custom(err.to_string())))
-}
+// Admin... keys are just signing keys w/ some different serialization; HcAc... instead of HcSc....
+#[derive(Debug)]
+pub struct AdminSigningPublicKey(pub SigningPublicKey);
```

Comment:
This is still an Ed25519 public key, so I don't think this should be a separate type.

Reply:
Right; I thought it made some sense to differentiate keys' "hcid" to indicate their semantics (eg. a `HcSc...` key is an agent signing key, while `HcAc...` is an Admin key.  But, I'm not convinced...